### PR TITLE
[CELEBORN-1711][TEST] Fix flaky test caused by master/worker setup issue

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?

1. retry on BindException when starting master/worker http server
2. record the used ports and pre-check whether the selected port is used or bounded before binding 

### Why are the changes needed?

To fix flaky test.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?
GA.

### CODE REVIEW

**Comment by turboFei** [comment]
```
[info] org.apache.celeborn.tests.spark.PushDataTimeoutTest *** ABORTED ***
[info]   java.lang.AssertionError: assertion failed
[info]   at scala.Predef$.assert(Predef.scala:208)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.$anonfun$setUpWorkers$7(MiniClusterFeature.scala:234)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.$anonfun$setUpWorkers$7$adapted(MiniClusterFeature.scala:234)
[info]   at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:149)
[info]   at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:237)
[info]   at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:230)
[info]   at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:44)
[info]   at scala.collection.mutable.HashMap.foreach(HashMap.scala:149)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpWorkers(MiniClusterFeature.scala:234)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpWorkers$(MiniClusterFeature.scala:181)
[info]   at org.apache.celeborn.tests.spark.PushDataTimeoutTest.setUpWorkers(PushDataTimeoutTest.scala:34)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpMiniCluster(MiniClusterFeature.scala:254)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpMiniCluster$(MiniClusterFeature.scala:250)
[info]   at org.apache.celeborn.tests.spark.PushDataTimeoutTest.setUpMiniCluster(PushDataTimeoutTest.scala:34)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setupMiniClusterWithRandomPorts(MiniClusterFeature.scala:77)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setupMiniClusterWithRandomPorts$(MiniClusterFeature.scala:49)
[info]   at org.apache.celeborn.tests.spark.PushDataTimeoutTest.setupMiniClusterWithRandomPorts(PushDataTimeoutTest.scala:34)
[info]   at org.apache.celeborn.tests.spark.PushDataTimeoutTest.beforeAll(PushDataTimeoutTest.scala:47)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:212)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.celeborn.tests.spark.PushDataTimeoutTest.run(PushDataTimeoutTest.scala:34)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:750)
```

**Comment by turboFei** [comment]
```
[info] org.apache.celeborn.tests.spark.SkewJoinSuite *** ABORTED ***
[info]   java.lang.AssertionError: assertion failed
[info]   at scala.Predef$.assert(Predef.scala:208)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.$anonfun$setUpWorkers$7(MiniClusterFeature.scala:234)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.$anonfun$setUpWorkers$7$adapted(MiniClusterFeature.scala:234)
[info]   at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:149)
[info]   at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:237)
[info]   at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:230)
[info]   at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:44)
[info]   at scala.collection.mutable.HashMap.foreach(HashMap.scala:149)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpWorkers(MiniClusterFeature.scala:234)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpWorkers$(MiniClusterFeature.scala:181)
[info]   at org.apache.celeborn.tests.spark.SkewJoinSuite.setUpWorkers(SkewJoinSuite.scala:32)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpMiniCluster(MiniClusterFeature.scala:254)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpMiniCluster$(MiniClusterFeature.scala:250)
[info]   at org.apache.celeborn.tests.spark.SkewJoinSuite.setUpMiniCluster(SkewJoinSuite.scala:32)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setupMiniClusterWithRandomPorts(MiniClusterFeature.scala:77)
[info]   at org.apache.celeborn.service.deploy.MiniClusterFeature.setupMiniClusterWithRandomPorts$(MiniClusterFeature.scala:49)
[info]   at org.apache.celeborn.tests.spark.SkewJoinSuite.setupMiniClusterWithRandomPorts(SkewJoinSuite.scala:32)
[info]   at org.apache.celeborn.tests.spark.SparkTestBase.beforeAll(SparkTestBase.scala:47)
[info]   at org.apache.celeborn.tests.spark.SparkTestBase.beforeAll$(SparkTestBase.scala:45)
[info]   at org.apache.celeborn.tests.spark.SkewJoinSuite.beforeAll(SkewJoinSuite.scala:32)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:212)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.celeborn.tests.spark.SkewJoinSuite.run(SkewJoinSuite.scala:32)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
```

**Comment by turboFei** [comment]
cc @SteNicholas @FMX @RexXiong 

The UT should be more stable after this pr.

**Comment by turboFei** [comment]
```
24/11/12 20:24:33,875 ERROR [ScalaTest-main-running-DiscoverySuite] ReusedExchangeSuite: cannot start all workers after 60000 ms
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:208)
	at org.apache.celeborn.service.deploy.MiniClusterFeature.$anonfun$setUpWorkers$7(MiniClusterFeature.scala:246)
	at org.apache.celeborn.service.deploy.MiniClusterFeature.$anonfun$setUpWorkers$7$adapted(MiniClusterFeature.scala:246)
	at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:149)
	at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:237)
	at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:230)
	at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:44)
	at scala.collection.mutable.HashMap.foreach(HashMap.scala:149)
	at org.apache.celeborn.service.deploy.MiniClusterFeature.setUpWorkers(MiniClusterFeature.scala:246)
```

**Comment by turboFei** [comment]
cc @FMX @RexXiong @SteNicholas It is to reduce the flaky tests.

**Comment by turboFei** [comment]
gentle ping @FMX @RexXiong @SteNicholas

**Comment by turboFei** [comment]
Seems the port conflicts issue happen with high frequency recently.

File: master/src/test/scala/org/apache/celeborn/service/deploy/master/MasterClusterFeature.scala [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    try {
+      socket.connect(new InetSocketAddress("localhost", port), 100)
+      true
+    } catch {
+      case _: IOException => false
+    } finally {
+      socket.close()
+    }
+  }
+  def selectRandomPort(): Int = {
```
Comment (onebox-li) [inline master/src/test/scala/org/apache/celeborn/service/deploy/master/MasterClusterFeature.scala:52]: Just a question, if we need sync here, seems workers are started in parallel.

Comment (turboFei) [reply master/src/test/scala/org/apache/celeborn/service/deploy/master/MasterClusterFeature.scala]: thanks, addressed

**Comment by turboFei** [comment]
interesting， so many failure recently .

**Comment by turboFei** [comment]
```
[info] - celeborn spark integration test - hash-checkDiskFull *** FAILED ***
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 6 in stage 8.0 failed 1 times, most recent failure: Lost task 6.0 in stage 8.0 (TID 24) (fv-az1146-454 executor driver): org.apache.celeborn.common.exception.CelebornIOException: Waiting timeout for task 4-6-0 while limiting zero in-flight requests
[info] 	at org.apache.celeborn.client.ShuffleClientImpl.limitZeroInFlight(ShuffleClientImpl.java:738)
[info] 	at org.apache.celeborn.client.ShuffleClientImpl.mapEndInternal(ShuffleClientImpl.java:1617)
[info] 	at org.apache.celeborn.client.ShuffleClientImpl.mapperEnd(ShuffleClientImpl.java:1601)
[info] 	at org.apache.spark.shuffle.celeborn.HashBasedShuffleWriter.close(HashBasedShuffleWriter.java:380)
[info] 	at org.apache.spark.shuffle.celeborn.HashBasedShuffleWriter.write(HashBasedShuffleWriter.java:176)
[info] 	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
[info] 	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:104)
[info] 	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
[info] 	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:166)
[info] 	at org.apache.spark.scheduler.Task.run(Task.scala:141)
[info] 	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
[info] 	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
[info] 	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
[info] 	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
[info] 	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[info] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[info] 	at java.base/java.lang.Thread.run(Thread.java:840)
[info] 
[info] Driver stacktrace:
[info]   at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2856)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2792)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2791)
[info]   at scala.collection.immutable.List.foreach(List.scala:333)
[info]   at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2791)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1247)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1247)
[info]   at scala.Option.foreach(Option.scala:437)
[info]   at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1247)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3060)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2994)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2983)
[info]   at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
[info]   at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:989)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2393)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2414)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2433)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2458)
[info]   at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1049)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
[info]   at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
[info]   at org.apache.spark.rdd.RDD.collect(RDD.scala:1048)
[info]   at org.apache.spark.rdd.PairRDDFunctions.$anonfun$collectAsMap$1(PairRDDFunctions.scala:738)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
[info]   at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
[info]   at org.apache.spark.rdd.PairRDDFunctions.collectAsMap(PairRDDFunctions.scala:737)
[info]   at org.apache.celeborn.tests.spark.SparkTestBase.repartition(SparkTestBase.scala:83)
[info]   at org.apache.celeborn.tests.spark.SparkTestBase.repartition$(SparkTestBase.scala:81)
[info]   at org.apache.celeborn.tests.spark.CelebornHashCheckDiskSuite.repartition(CelebornHashCheckDiskSuite.scala:32)
[info]   at org.apache.celeborn.tests.spark.CelebornHashCheckDiskSuite.$anonfun$new$1(CelebornHashCheckDiskSuite.scala:70)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[info]   at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
[info]   at org.apache.celeborn.tests.spark.CelebornHashCheckDiskSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(CelebornHashCheckDiskSuite.scala:32)
[info]   at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
[info]   at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
[info]   at org.apache.celeborn.tests.spark.CelebornHashCheckDiskSuite.runTest(CelebornHashCheckDiskSuite.scala:32)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:333)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
[info]   at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1564)
[info]   at org.scalatest.Suite.run(Suite.scala:1114)
[info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[info]   at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1564)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
[info]   at org.apache.celeborn.tests.spark.CelebornHashCheckDiskSuite.org$scalatest$BeforeAndAfterAll$$super$run(CelebornHashCheckDiskSuite.scala:32)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.celeborn.tests.spark.CelebornHashCheckDiskSuite.run(CelebornHashCheckDiskSuite.scala:32)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
[info]   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[info]   at java.base/java.lang.Thread.run(Thread.java:840)
[info]   Cause: org.apache.celeborn.common.exception.CelebornIOException: Waiting timeout for task 4-6-0 while limiting zero in-flight requests
[info]   at org.apache.celeborn.client.ShuffleClientImpl.limitZeroInFlight(ShuffleClientImpl.java:738)
[info]   at org.apache.celeborn.client.ShuffleClientImpl.mapEndInternal(ShuffleClientImpl.java:1617)
[info]   at org.apache.celeborn.client.ShuffleClientImpl.mapperEnd(ShuffleClientImpl.java:1601)
[info]   at org.apache.spark.shuffle.celeborn.HashBasedShuffleWriter.close(HashBasedShuffleWriter.java:380)
[info]   at org.apache.spark.shuffle.celeborn.HashBasedShuffleWriter.write(HashBasedShuffleWriter.java:176)
[info]   at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
[info]   at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:104)
[info]   at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
[info]   at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:166)
[info]   at org.apache.spark.scheduler.Task.run(Task.scala:141)
[info]   at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
[info]   at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
[info]   at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
[info]   at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
[info]   at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[info]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[info]   at java.base/java.lang.Thread.run(Thread.java:840)
```

**Review by onebox-li — APPROVED** [review]

**Comment by turboFei** [comment]
There is only change for testing code. It should be fine.
If no more comments, I plan to merge it tomorrow.

cc @RexXiong

**Review by RexXiong — APPROVED** [review]
LGTM

**Pull request closed without merging**